crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -304,110 +296,319 @@ where
-    let mut unepoched_tokens = HashMap::new();
+    tokens: &mut BTreeMap<
+        AssetType,
+        (Address, token::Denomination, MaspDigitPos),
+    >,
+) -> Result<()> {
     for digit in MaspDigitPos::iter() {
         let asset_type = encode_asset_type(token.clone(), denom, digit, None)
             .wrap_err("unable to create asset type")?;
-        unepoched_tokens.insert(asset_type, (token.clone(), denom, digit));
+        tokens.insert(asset_type, (token.clone(), denom, digit));
+    }
+    Ok(())
+}
+
+// Handle transparent input
+fn validate_transparent_input<A: Authorization>(
+    vin: &TxIn<A>,
+    changed_balances: &mut ChangedBalances,
+    transparent_tx_pool: &mut I128Sum,
+    epoch: Epoch,
+    conversion_state: &ConversionState,
+    signers: &mut BTreeSet<TransparentAddress>,
+) -> Result<()> {
+    // A decrease in the balance of an account needs to be
+    // authorized by the account of this transparent input
+    signers.insert(vin.address);
+    // Non-masp sources add to the transparent tx pool
+    *transparent_tx_pool = transparent_tx_pool
+        .checked_add(
+            &I128Sum::from_nonnegative(vin.asset_type, vin.value as i128)
+                .ok()
+                .ok_or_err_msg("invalid value or asset type for amount")?,
+        )
+        .ok_or_err_msg("Overflow in input sum")?;
+
+    let bal_ref = changed_balances
+        .pre
+        .entry(vin.address)
+        .or_insert(ValueSum::zero());
+
+    match conversion_state.assets.get(&vin.asset_type) {
+        // Note how the asset's epoch must be equal to the present: users
+        // must never be allowed to backdate transparent inputs to a
+        // transaction for they would then be able to claim rewards while
+        // locking their assets for negligible time periods.
+        Some(((address, _asset_denom, digit), asset_epoch, _, _))
+            if *asset_epoch == epoch =>
+        {
+            let amount =
+                token::Amount::from_masp_denominated(vin.value, *digit);
+            *bal_ref = bal_ref
+                .checked_sub(&ValueSum::from_pair(address.clone(), amount))
+                .ok_or_else(|| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in bundle balance",
+                    ))
+                })?;
+        }
+        // Maybe the asset type has no attached epoch
+        None if changed_balances.tokens.contains_key(&vin.asset_type) => {
+            let (token, denom, digit) =
+                &changed_balances.tokens[&vin.asset_type];
+            // Determine what the asset type would be if it were epoched
+            let epoched_asset_type =
+                encode_asset_type(token.clone(), *denom, *digit, Some(epoch))
+                    .wrap_err("unable to create asset type")?;
+            if conversion_state.assets.contains_key(&epoched_asset_type) {
```

Comment:
This check is here to prevent an unepoched asset from entering the shielded pool in the case where its underlying token address is participating in the shielded rewards program. But it is possible for an unepoched asset to enter the shielded pool and then later have its underlying token address participate in the shielded pool. In this case we need to allow users to withdraw these assets even though their epoched equivalents are in the conversion state. This was my reasoning for why this check is not repeated in the context of the transparent output. But please let me know if I'm missing something!